docs/madr/1-reassign-evolve.md

```diff
@@ -0,0 +1,166 @@
+
+With the ARIO process now supporting
+[`Reassign-Name`](https://github.com/ar-io/ar-io-ant-process/pull/26) and recent
+additions to add [`_boot`](https://github.com/ar-io/ar-io-ant-process/pull/57)
+to fork the process state to a whole new ANT, we can assign that new ANT as the
+registered ANT to the specified ArNS name in the registry, and using the module
+ID of the process to identify process version and capabilities, and solely the
+module ID.
+
+This optimized approach reduces computational overhead and improves caching for
+better performance. It also shifts to a more robust upgrade mechanism using
+immutable module IDs rather than version detection heuristics. We can do this
+now that we compile and publish our own WASM binaries, rather than using an AOS
+binary and loading our code into it.
+
+```mermaid
+sequenceDiagram
+    autonumber
+
+    participant ARIOGateway
+    participant Owner
+    participant OldAnt
+    participant NewAnt
+    participant AntRegistry
+    participant ARIOProcess
+
+
+    Owner ->> ARIOGateway: Get Module ID for process from GQL
+    ARIOGateway -->> Owner: Module ID Result
+
+
+    alt No upgrade needed Module ID from process meta is up to date
+
+        else
+
+        Owner->>OldAnt: Get state
+        OldAnt-->>Owner: Return state
+
+        Owner->>NewAnt: Spawn new ant with old state
+        NewAnt->>AntRegistry: Register ANT notice
+        Owner->>NewAnt: Get State to verify responsiveness and Module ID to verify correct deployment
+
+        Owner->>OldAnt: Sign and accept Reassign message
+        OldAnt->>ARIOProcess: Forward Reassign Message
+        ARIOProcess-->>OldAnt: Reassign-Notice
+
+        Owner->>ARIOProcess: Check for Reassignment success and notify user
+
+
+    end
+```
+
+### Positive Consequences
+
+- **Performance Improvement:** Reducing dry runs significantly speeds up ANT
+  loading.
+- **Immutable Module Versioning:** Eliminates issues with heuristic version
+  analysis.
+  - Downstream clients can identify and validate the capabilities of a module
+    (and by extension the ANT itself) has by maintaining a map of modules that
+    are capables of executing target workflows
+- **Reduced CU Load:** Fewer computations per ANT.
+- **Gateway ANT Module Whitelisting:** Gateways can decide which ANT Modules
+  they support.
+
+### Negative Consequences
+
+- **Complicates Memory Usage on ANT Registry:** More ANTs means more state in
+  the ANT registry, which currently doesn't have a cleanup process.
+- **Potential Integration Issues:** Existing integrations relying on fixed ANT
+  IDs may break.
+- **State Limitations:** Large ANT states (e.g., >2000 undernames) may fail to
+  bootstrap on spawn, meaning we need to understand the limitations there and
+  decide what amounts we wish to support.
+- **MU cranking risks**: Its possible a module fails to instantiate/boot
+  properly - the defense there is to query for its state to ensure
+  responsiveness
```

Comment:
once spawned. Users will not be presented to reassign a name if spawning fails, or a newly created process is not as expected.

Reply:
[e8606a0](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/e8606a049e2e40d833911437d53af117306c13d9)